spirv_glsl.cpp

```diff
@@ -3258,6 +3306,78 @@ void CompilerGLSL::emit_resources()
+	{
+		statement("#ifdef GL_ARB_shader_group_vote");
+		statement("bool subgroupAll(in bool v) { return allInvocationsARB(v); }");
+		statement("bool subgroupAny(in bool v) { return anyInvocationARB(v); }");
+		statement("#define SUBGROUP_ALL_EQUAL_WORKAROUND(type) bool subgroupAllEqual(type value) { return "
+		          "allInvocationsEqualARB(equal(value)); }");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(int)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uint)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(float)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(double)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec4)");
+		statement("#undef SUBGROUP_ALL_EQUAL_WORKAROUND");
+
+		statement("#elif defined(GL_NV_gpu_shader5)");
+		statement("bool subgroupAll(in bool value) { return allThreadsNV(value); }");
+		statement("bool subgroupAny(in bool value) { return anyThreadNV(value); }");
+		statement("bool subgroupAllEqual(in bool value) { return allThreadsEqualNV(value); }");
+
+		statement("#elif defined(GL_AMD_gcn_shader)&&defined(GL_AMD_gpu_shader_int64)");
+		statement("bool subgroupAll(in bool value) { return ballotAMD(value)==ballotAMD(true); }");
+		statement("bool subgroupAny(in bool value) { return ballotAMD(value)!=0ull; }");
+		statement("bool subgroupAllEqual(in bool value) { uint64_t b=ballotAMD(value); return b==0uLL || "
+		          "b==ballotAMD(true); }");
+
+		statement("#endif");
+	}
+	//GL_KHR_shader_subgroup_basic
+	if (!options.vulkan_semantics && declaredKHR_shader_subgroup_basic)
+	{
+	/* TODO?
+	https://github.com/KhronosGroup/SPIRV-Cross/issues/1351
+	Neither GL_ARB_shader_ballot, GL_NV_shader_thread_group or AMD_gcn_shader support subgroup barriers (execution or memory) explicitly.
+	However the wording of GL_NV_shader_thread_group and ARB_shader_ballot explicitly stat that invocations run in lock-step.
+	So this should mean that Turing and Volta (or any other GPU architecture which supports per-invocation Program Counters)
+	should either not report these extensions or insert synchronization implicitly in the shaders using those extensions.
+	**This could allow us to turn any subgroupBarrier into a NO-OP.**
+	However the extensions do not guarantee anything about memory dependencies.
+	So the presence of subgroupMemoryBarrier* function call in the SPV should produce an error.
+	*/
```

Comment:
yep, error out if `subgroupMemoryBarrier*` SPIR-V op is attempted to be emitted.

But turn the `subgroupBarrier` into a NOOP (maybe emit a comment for clarity).

Reply:
> But turn the subgroupBarrier into a NOOP (maybe emit a comment for clarity).

Doesnt matter in case when error is generated anyway

Reply:
having a `subgroupBarrier` shouldn't be an error, its an execution barrier which with the ARB, AMD and NV extensions we're using turns into a NOOP (because the extensions state that subgroups execute in lockstep so this barrier is implicit).